kimchi/src/tests/foreign_field_mul.rs

```diff
@@ -927,6 +927,57 @@ fn test_invalid_carry1_hi_plookup() {
     );
 }
 
+#[test]
+fn test_invalid_wraparound_carry1_hi_plookup() {
+    let a = BigUint::zero();
+    let b = BigUint::zero();
+
+    // Sanity check wraparound values
+    let two_to_9 = PallasField::from(2u32).pow(&[9]);
+    // Wraparound (exploit) value x s.t. x >= 2^12 AND 2^9 * x < 2^12
+    // (credit to querolita for computing the real instances of this value for these test cases!)
+    let wraparound_0 = two_to_9.inverse().expect("failed to get inverse");
+    for i in 0..8 {
+        let wraparound_i = wraparound_0 + PallasField::from(i);
+        assert!(wraparound_i >= PallasField::from(2u32).pow(&[12u64]));
+        assert!(two_to_9 * wraparound_i < PallasField::from(2u32).pow(&[12u64])); // Wraparound!!!
+    }
+    // edge case: x - 1 is not a wraparound value
+    assert!(wraparound_0 - PallasField::one() >= PallasField::from(2u32).pow(&[12u64]));
+    assert!(two_to_9 * (wraparound_0 - PallasField::one()) >= PallasField::from(2u32).pow(&[12u64]));
+    // edge case: x + 8 is not a wraparound value
+    assert!(wraparound_0 + PallasField::from(8) >= PallasField::from(2u32).pow(&[12u64]));
+    assert!(two_to_9 * (wraparound_0 + PallasField::from(8)) >= PallasField::from(2u32).pow(&[12u64]));
+
+    // Invalid carry1_hi witness that causes wrap around to something less than 3-bits
+    let (result, witness) = run_test::<Vesta, VestaBaseSponge, VestaScalarSponge>(
+        true,
+        false, // Disable external checks so we can catch carry1_hi plookup failure
```

Comment:
No need now, but to make the code self-documenting, we really should either
* make this function call take a record
* create a dedicated enum `CheckConstraints::{WithGateConstraints, WithoutGateConstraints}`

Reply:
It would also be good to
* make it only create CS once and then test many witnesses against it
* factor it out so could be used by other tests